kore/src/Kore/Step/Axiom/EvaluationStrategy.hs

```diff
@@ -92,6 +91,64 @@ definitionEvaluation rules =
     BuiltinAndAxiomSimplifier
         (evaluateWithDefinitionAxioms rules)
 
+{- | Creates an evaluator for a function all the rules that define it.
+
+The function is not applied (@totalDefinitionEvaluation@ returns
+'AttemptedAxiom.NotApplicable') if the supplied rules do not match the entire
+input.
+
+See also: 'definitionEvaluation'
+
+-}
+totalDefinitionEvaluation
+    :: forall level
+    .  [EqualityRule level Variable]
+    -> BuiltinAndAxiomSimplifier level
+totalDefinitionEvaluation rules =
+    BuiltinAndAxiomSimplifier totalDefinitionEvaluationWorker
+  where
+    totalDefinitionEvaluationWorker
+        ::  forall variable
+        .   ( FreshVariable variable
+            , MetaOrObject level
+            , Ord (variable level)
+            , OrdMetaOrObject variable
+            , SortedVariable variable
+            , Show (variable level)
+            , Show (variable Object)
+            , Unparse (variable level)
+            , ShowMetaOrObject variable
+            )
+        => MetadataTools level StepperAttributes
+        -> PredicateSubstitutionSimplifier level
+        -> StepPatternSimplifier level
+        -> BuiltinAndAxiomSimplifierMap level
+        -> StepPattern level variable
+        -> Simplifier
+            ( AttemptedAxiom level variable
+            , SimplificationProof level
+            )
+    totalDefinitionEvaluationWorker
+        tools
+        predicateSimplifier
+        termSimplifier
+        axiomSimplifiers
+        term
+      = do
+        (result, proof) <- evaluate
+        if AttemptedAxiom.hasRemainders result
+            then return (AttemptedAxiom.NotApplicable, proof)
+            else return (result, proof)
+      where
+        evaluate =
+            evaluateWithDefinitionAxioms
+                rules
+                tools
+                predicateSimplifier
+                termSimplifier
+                axiomSimplifiers
+                term
```

Comment:
Why not leave 'term' as an argument to 'evaluate'? You did something similar in most of the last PRs I reviewed, and I think that the `evaluate` (or equivalent) call looked better that way.